agent/consul/stream/event_publisher.go

```diff
@@ -0,0 +1,305 @@
+	if len(events) > 0 {
+		e.publishCh <- changeEvents{events: events}
+	}
+}
+
+func (e *EventPublisher) handleUpdates(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			// TODO: also close all subscriptions so the subscribers are moved
+			// to the new publisher?
+			return
+		case update := <-e.publishCh:
+			e.sendEvents(update)
+		}
+	}
+}
+
+// sendEvents sends the given events to any applicable topic listeners, as well
+// as any ACL update events to cause affected listeners to reset their stream.
+func (e *EventPublisher) sendEvents(update changeEvents) {
+	for _, event := range update.events {
+		if unsubEvent, ok := event.Payload.(closeSubscriptionPayload); ok {
+			e.subscriptions.closeSubscriptionsForTokens(unsubEvent.tokensSecretIDs)
+		}
+	}
+
+	eventsByTopic := make(map[Topic][]Event)
+	for _, event := range update.events {
+		eventsByTopic[event.Topic] = append(eventsByTopic[event.Topic], event)
+	}
+
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	for topic, events := range eventsByTopic {
+		e.getTopicBuffer(topic).Append(events)
+	}
+}
+
+// getTopicBuffer for the topic. Creates a new event buffer if one does not
+// already exist.
+//
+// EventPublisher.lock must be held to call this method.
+func (e *EventPublisher) getTopicBuffer(topic Topic) *eventBuffer {
+	buf, ok := e.topicBuffers[topic]
+	if !ok {
+		buf = newEventBuffer()
+		e.topicBuffers[topic] = buf
+	}
+	return buf
+}
+
+// Subscribe returns a new stream.Subscription for the given request. A
+// subscription will stream an initial snapshot of events matching the request
+// if required and then block until new events that modify the request occur, or
+// the context is cancelled. Subscriptions may be forced to reset if the server
+// decides it can no longer maintain correct operation for example if ACL
+// policies changed or the state store was restored.
+//
+// When the caller is finished with the subscription for any reason, it must
+// call Subscription.Unsubscribe to free ACL tracking resources.
+func (e *EventPublisher) Subscribe(
+	ctx context.Context,
+	req *SubscribeRequest,
+) (*Subscription, error) {
+	// Ensure we know how to make a snapshot for this topic
+	_, ok := e.snapshotHandlers[req.Topic]
+	if !ok {
+		return nil, fmt.Errorf("unknown topic %v", req.Topic)
+	}
+
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// Ensure there is a topic buffer for that topic so we start capturing any
+	// future published events.
+	buf := e.getTopicBuffer(req.Topic)
+
+	// See if we need a snapshot
+	topicHead := buf.Head()
+	var sub *Subscription
+	if req.Index > 0 && len(topicHead.Events) > 0 && topicHead.Events[0].Index == req.Index {
+		// No need for a snapshot, send the "resume stream" message to signal to
+		// client it's cache is still good. (note that this can be distinguished
+		// from a legitimate empty snapshot due to the index matching the one the
+		// client sent), then follow along from here in the topic.
```

Comment:
```suggestion
		// client its cache is still good, then follow along from here in the topic.
```
This was out of date and predated a distinct Resume message.